File: ce326/hw2/PPMImage.java

package ce326.hw2;

import java.io.FileNotFoundException;
import java.io.File;
import ce326.hw2.UnsupportedFileFormatException;
import java.util.Scanner;
import java.io.File;  
import java.io.IOException; 
import java.io.FileWriter; 
import java.util.Scanner;
import java.io.FileReader;  
import java.io.PrintWriter; 
import java.util.ArrayList;
import java.io.FileInputStream; 
import java.nio.file.FileAlreadyExistsException;
import java.util.logging.Level;
import java.util.logging.Logger;
import java.util.Formatter;
import java.io.*;  
import java.lang.*;
import static java.lang.Character.isWhitespace;

public class PPMImage extends RGBImage{
    File f;
    
    public PPMImage(java.io.File file) throws FileNotFoundException, UnsupportedFileFormatException{//a copy of a ppm image
        int times = 0;//mas leei se poia metavliti anaferetai h kathe timi pou yparxei sto arxeio
        int first = 1;//distinguish height, width, colordepth from red, blue, green
        int heightHelp = 0;
        int widthHelp = 0;
        String strPpm;
        
        try{
            strPpm = file.getName();
            if(!((strPpm.charAt(strPpm.length() - 1) == 'm') & (strPpm.charAt(strPpm.length() - 2) == 'p') & (strPpm.charAt(strPpm.length() - 3) == 'p') & (strPpm.charAt(strPpm.length() - 4) == '.'))){
                throw new UnsupportedFileFormatException();
            }
            
            Scanner readerFile = new Scanner(file);
            readerFile.next();//traverse first object(P3)

            while(readerFile.hasNext()){
                times++;
                if(first == 1){
                    if(times == 1){
                        width = Integer.parseInt(readerFile.next());
                    }
                    else if(times == 2){
                        height = Integer.parseInt(readerFile.next());
                    }
                    else if(times == 3){
                        colordepth = Integer.parseInt(readerFile.next());

                         array = new RGBPixel[height][width];
                                for(int k = 0; k < height; k++){//initialisation rgb array
                                    for(int j = 0; j < width; j++){
                                        array[k][j] = new RGBPixel();
                                    }
                                }

                        first = 0;
                        times = 0;
                    }
                }
                else{
                   if(times == 1){//times = 1 ->red colour
                        array[heightHelp][widthHelp].Red = (short) Integer.parseInt(readerFile.next());
                    }
                    else if(times == 2){//times = 2 ->green colour
                         array[heightHelp][widthHelp].Green = (short) Integer.parseInt(readerFile.next());
                    }
                    else if(times == 3){//times = 3 ->blue colour
                        array[heightHelp][widthHelp].Blue = (short) Integer.parseInt(readerFile.next());
                        times = 0;//initialisation of height/width for next loop
                         if((heightHelp == height-1) & (widthHelp == width-1)){
                            break;
                        }
                        else if(widthHelp == (width-1)){//reinitialise so that colours get in the right pixels
                            widthHelp = 0;
                            heightHelp++;
                        }
                        else{
                            widthHelp++;
                        }
                    } 

                }

            }
        }catch(FileNotFoundException e){
            
        }catch(UnsupportedFileFormatException ex){
            
        }
     
    }

    public PPMImage(RGBImage img){        
        this.height = img.height;
        this.width = img.width;
        this.colordepth = img.colordepth;
        this.array = new RGBPixel[height][width];
        
        for(int i = 0; i < height; i++){
            for(int j = 0; j < width; j++){
                this.array[i][j]= img.array[i][j];
            }
        }
    }
    
    public PPMImage(){//dhmiourgia epipleon kataskeyasti gia arxikopoiiseis se alles klaseis
        
    }

    public PPMImage(YUVImage img){
        this.height = img.height;
        this.width = img.width;
        this.colordepth = 255;
        this.array = new RGBPixel[height][width];
        
        for(int i = 0; i < height; i++){
            for(int j = 0; j < width; j++){
                this.array[i][j] = new RGBPixel(img.array[i][j]);
            }
        }
    
    }
    
    
    //initilisation of an image created from ppmimagestacker
    public PPMImage(RGBImage img, int number){
        PrintWriter pw = null;
        System.out.println("mpika mesa stin ppmimage");
        
        this.height = img.height;
        this.width = img.width;
        this.colordepth = img.colordepth;
        this.array = new RGBPixel[height][width];
     }
    
    @Override
    public String toString(){
        String str = "";
        StringBuilder strbuilderHelp = new StringBuilder(str);
        strbuilderHelp.insert(strbuilderHelp.length(), "P3");
        strbuilderHelp.insert(strbuilderHelp.length(), String.format(" %d %d %d", width, height, colordepth));
        
       for(int i = 0; i < height; i++){
           for(int j = 0; j < width; j++){
               strbuilderHelp.insert(strbuilderHelp.length(), String.format(" %d %d %d", array[i][j].Red, array[i][j].Green, array[i][j].Blue));
           }
       }  
       
       str = String.valueOf(strbuilderHelp);
       System.out.println(str);
            
        return(str);
    }
    
    public void toFile(java.io.File file){
        PrintWriter pw = null;
        
        try{
            this.f = file;
            
            if(file.exists()){//if the folder alredy existed -> empty
                PrintWriter print_writer = new PrintWriter(file);
                print_writer.print("");
                print_writer.close();
            }
            
            FileWriter writer = new FileWriter(file, true);
            pw = new PrintWriter(file);
            pw.println("P3" + "\n" + width + " " + height + " " + "255");
            for(int i = 0; i < height; i++){
                for(int j = 0; j <width; j++){
                    pw.println(array[i][j].Red + " " + array[i][j].Green + " " + array[i][j].Blue);
                }
        }
        }catch (IOException e) {
            System.out.println("An error occurred during reading");
        }finally {
            if (pw != null) {
               pw.close();
            }
        }
        
    }      
}

File: ce326/hw2/python/PPMImage.py
from RGBImage import RGBImage
from RGBPixel import RGBPixel
from UnsupportedFileFormatException import UnsupportedFileFormatException
import os
import numpy as np
import os.path
import io
import imghdr
from os.path import exists
from os import path

class PPMImage(RGBImage):
        
    def __init__(self, f): #throws FileNotFoundException, UnsupportedFileFormatException//a copy of a ppm image
           
        #if f is an image
        if(isinstance(f, RGBImage)):
            self.height = f.height;
            self.width = f.width;
            self.colordepth = f.colordepth;
            self.array = np.empty(shape = (self.width, self.height), dtype = RGBPixel);
            
            for i in range(self.height):
                for j in range(self.width):
                    self.array[i][j]= f.array[i][j];
        
        #if the f is a file
        elif(isinstance(f, io.IOBase)):
            times = 0; #mas leei se poia metavliti anaferetai h kathe timi pou yparxei sto arxeio
            first = 1; #distinguish height, width, colordepth from red, blue, green
            heightHelp = 0;
            widthHelp = 0;
            var = 0;
            
            self.f = f;
            
            try:
                strPpm = os.path.basename(f.name)
                if((strPpm[len(strPpm) - 4 : len(strPpm)]) != '.ppm'):
                    raise UnsupportedFileFormatException()
                
                string = f.read().rstrip('\n')
       
                splits = string.split(); #traverse first object(P3)
                
                #read the information from the .ppm file and save them to array, width, height, colordepth
                for split in splits:
                    var = var + 1
                    if(var == 1):
                        continue
                    
                    times = times + 1;
                    if first == 1:
                        if times == 1:
                            self.width = int(split);
                        elif times == 2:
                            self.height = int(split);
                        elif times == 3:
                            self.colordepth = int(split);
                            
                            self.array = np.empty(shape = (self.height, self.width), dtype = RGBPixel)
                            for k in range(self.height): #initialisation rgb array
                                for j in range(self.width):
                                    self.array[k][j] = RGBPixel(0, 0, 0, 255);
                                    
                            first = 0;
                            times = 0;
        
                    else:
                       if times == 1: #times = 1 ->red colour
                            self.array[heightHelp][widthHelp].Red = int(split);
        
                       elif times == 2: #times = 2 ->green colour
                            self.array[heightHelp][widthHelp].Green = int(split);
                        
                       elif times == 3: #times = 3 ->blue colour
                            self.array[heightHelp][widthHelp].Blue = int(split);
                            times = 0; #initialisation of height/width for next loop
                            
                            if (heightHelp == self.height - 1) & (widthHelp == self.width - 1):
                                break;
                                
                            elif widthHelp == (self.width - 1): #reinitialise so that colours get in the right pixels
                                widthHelp = 0;
                                heightHelp = heightHelp + 1;
                                
                            else:
                                widthHelp = widthHelp + 1;
                            
            except FileNotFoundError:
                print("File not Found!")
            except UnsupportedFileFormatException:
                print("UnsupportedFileFormatException")
                
                
        else:
            self.height = 0;
            self.width = 0;
            self.colordepth = 0;
            self.array = np.empty(shape = (self.width, self.height), dtype = RGBPixel);            
                
    
    
    #initilisation of an image created from ppmimagestacker
    def PPMImage_fromstacker(self, img, number):
        self.height = img.height;
        self.width = img.width;
        self.colordepth = img.colordepth;
        self.array = np.empty(shape = (self.width, self.height), dtype = RGBPixel);
    
    def toString(self):        
        mystr = "P3\n"
        mystr += str(self.width);
        mystr += " "
        mystr += str(self.height);
        mystr += "\n"
        mystr += str(self.colordepth);
        mystr += "\n"
        
        for i in range(self.height):
            for j in range (self.width):
                 mystr += str(self.array[i][j].Red)
                 mystr += " "
                 mystr += str(self.array[i][j].Green)
                 mystr += " "
                 mystr += str(self.array[i][j].Blue)
                 mystr += " "
                 
            mystr += "\n"
        
        return(mystr)
    
    def toFile(self, file):
        try:
            self.f = file
            
            #delete the contents of the file
            file.truncate(0)
            
            string = "P3";
            string += "\n";
            string += str(self.width);
            string += " ";
            string += str(self.height);
            string += "\n";
            string += "255";
            string += "\n";
        
            for i in range(self.height):
                for j in range(self.width):
                    string += str(self.array[i][j].Red)
                    string += " ";
                    string += str(self.array[i][j].Green)
                    string += " ";
                    string += str(self.array[i][j].Blue)
                    string += " ";
                string += "\n";
            
            file.write(string)
        
        except IOError:
            print("An error occurred during writing");
    
